# ข้อ 1 — พนักงานส่งพัสดุในเมือง (TSP + Knapsack)

## ข้อ1.1 TSP

In [ ]:
from itertools import permutations  #itertools เป็น Module ทีมีเครื่องมือช่วยจัดการข้อมูลการวนซ้ำ || itertools = กล่องเครื่องมือที่ Python เตรียมไว้ให้ สำหรับสร้าง/จัดการลำดับข้อมูลต่าง ๆ
# permutations การสับเปลี่ยนโดยสนใจลำดับ
# ระยะทางระหว่างจุด
distance = {
    'A': {'B': 4, 'C': 7, 'D': 9, 'E': 5},
    'B': {'A': 4, 'C': 3, 'D': 6, 'E': 8},
    'C': {'A': 7, 'B': 3, 'D': 2, 'E': 6},
    'D': {'A': 9, 'B': 6, 'C': 2, 'E': 4},
    'E': {'A': 5, 'B': 8, 'C': 6, 'D': 4}
}

def tsp_exhaustive(points):
    best_route = None
    best_distance = float('inf')

    # ลองทุกลำดับของจุด
    for route in permutations(points):
        total_distance = 0

        # A → จุดแรก
        total_distance += distance['A'][route[0]]

        # เดินจากจุดหนึ่งไปอีกจุด
        for i in range(len(route) - 1):
            total_distance += distance[route[i]][route[i + 1]]

        # จุดสุดท้าย → A
        total_distance += distance[route[-1]]['A']

        # เก็บเส้นทางที่สั้นที่สุด
        if total_distance < best_distance:
            best_distance = total_distance
            best_route = route

    return best_route, best_distance


points = ['B', 'C', 'D', 'E']

route, total = tsp_exhaustive(points)

print("เส้นทางที่สั้นที่สุด:")
print("A →", " → ".join(route), "→ A")
print("ระยะทางรวม:", total)

Best route: ('A', 'B', 'C', 'D', 'E', 'A')
Distance: 18


## ข้อ1.2 Knapsack 

In [11]:
from itertools import combinations
# permutations การจัดหมู่ แบบไใาว
packages = [
    ("P1", 30, 400),
    ("P2", 25, 700),
    ("P3", 20, 900),
    ("P4", 35, 500),
    ("P5", 15, 400)
]

capacity = 80
best_packages = None
best_weight = 0
best_value = 0

# ลองทุกชุดของพัสดุ
for r in range(len(packages) + 1):
    for selected in combinations(packages, r):
        total_weight = sum(item[1] for item in selected)
        total_value = sum(item[2] for item in selected)

        # ต้องไม่เกิน 80 kg
        if total_weight <= capacity:
            if total_value > best_value:
                best_packages = selected
                best_weight = total_weight
                best_value = total_value


print("พัสดุที่เลือก:")

for package in best_packages:
    print(package[0])

print("น้ำหนักรวม:", best_weight, "kg")
print("ค่าส่งรวม:", best_value, "บาท")

พัสดุที่เลือก:
P2
P3
P4
น้ำหนักรวม: 80 kg
ค่าส่งรวม: 2100 บาท


1.3 วิเคราะห์ว่าท าไมวิธี Brute Force จึงไม่เหมาะกับกรณี 20 จุด : 

Brute Force ไม่เหมาะกับ 20 จุด เพราะจำนวนเส้นทางโตแบบ Factorial: 20! = 2, 432, 902, 008, 176, 640, 000 จึงใช้เวลามหาศาลเมื่อ n เพิ่มขึ้น

# ข้อ 2 จัดตารางเวรพยาบาล (Assignment Problem)

In [7]:
from itertools import permutations

# คะแนนของพยาบาลแต่ละคนในแต่ละกะ
score = [
    [3, 5, 7, 2],  # นิด
    [6, 2, 4, 8],  # หน่อย
    [4, 7, 3, 5],  # นุ่น
    [2, 6, 5, 4]   # เนย
]

nurses = ["นิด", "หน่อย", "นุ่น", "เนย"]
shifts = ["เช้า", "บ่าย", "ดึก", "พัก"]

best_score = float("inf")
best_assignment = None

# ลองจัดกะทุกแบบ
for assignment in permutations(range(4)):

    total = 0

    for i in range(4):
        total += score[i][assignment[i]]

    # คะแนนน้อยที่สุด = ดีที่สุด
    if total < best_score:
        best_score = total
        best_assignment = assignment


# แสดงผล
print("ผลการจัดกะที่ดีที่สุด")

for i in range(4):
    print(nurses[i], "->", shifts[best_assignment[i]])

print("คะแนนรวม =", best_score)

ผลการจัดกะที่ดีที่สุด
นิด -> พัก
หน่อย -> บ่าย
นุ่น -> ดึก
เนย -> เช้า
คะแนนรวม = 9


2.3 หากมีพยาบาลเพิ่มเป็น 10 คน จ านวนกรณีที่ต้องตรวจจะเป็นเท่าใด? และ O(n!) หมายความว่า อย่างไรในทางปฏิบัติ?

ถ้ามีพยาบาล 10 คน: 10! = 3,628,800 วิธี O(n!) หมายถึงจำนวนกรณีเพิ่มขึ้นแบบ Factorial อย่างรวดเร็ว ทำให้เมื่อ n ใหญ่ขึ้น Exhaustive Search ใช้เวลามากจนไม่เหมาะกับงานจริง

# ข้อ 3 — วางต าแหน่งสถานีชาร์จรถไฟฟ้า (Closest-Pair + Convex Hull)

In [13]:
from math import sqrt

points = {
    "P1": (2, 3),
    "P2": (5, 8),
    "P3": (9, 2),
    "P4": (1, 7),
    "P5": (8, 9),
    "P6": (6, 4)
}

names = list(points)

best_distance = float("inf")
best_pair = None

for i in range(len(names)):
    for j in range(i + 1, len(names)):
        x1, y1 = points[names[i]]
        x2, y2 = points[names[j]]

        d = sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)

        if d < best_distance:
            best_distance = d
            best_pair = (names[i], names[j])

print("คู่ที่ใกล้ที่สุด =", best_pair)
print("ระยะห่าง =", best_distance)

คู่ที่ใกล้ที่สุด = ('P2', 'P5')
ระยะห่าง = 3.1622776601683795


3.1 จำนวนคู่: C(6,2) = 6×5/2 = 15 คู่

คู่ที่ใกล้ที่สุดคือ: P2 และ P5

ระยะ: sqrt((8-5)^2 + (9-8)^2)= sqrt(10)

ดังนั้น minimum distance = sqrt(10)

ตรวจคู่สำคัญ:
P1-P6 = sqrt(17) 

P2-P5 = sqrt(10) 

P3-P6 = sqrt(13) 

P4-P2 = 5 

P5-P6 = sqrt(26)

ดังนั้น P2-P5 ใกล้ที่สุด

3.2 จุดที่อยู่ขอบนอกคือ: P4 → P1 → P3 → P5 → P4 จุด P2 และ P6 อยู่ด้านใน

ดังนั้นขอบ Hull คือ: 
P4-P1

P1-P3

P3-P5

P5-P4

3.3 ถ้ามี 1,000 จุด: C(1000,2)= 1000×999/2= 499,500 คู่

Brute Force มี Time Complexity O(n²) เพราะจำนวนคู่เพิ่มประมาณ n²


# ข้อ 4 วางแผนลงทะเบียนเรียน (Topological Sort + Insertion Sort)

In [14]:
from itertools import permutations

nodes = ["V1", "V2", "V3", "V4", "V5", "V6", "V7"]

edges = [
    ("V1", "V2"),
    ("V1", "V3"),
    ("V2", "V4"),
    ("V3", "V4"),
    ("V3", "V5"),
    ("V4", "V6"),
    ("V5", "V6"),
    ("V6", "V7")
]

valid_orders = []

for order in permutations(nodes):
    position = {node: i for i, node in enumerate(order)}

    if all(position[a] < position[b] for a, b in edges):
        valid_orders.append(order)

for order in valid_orders:
    print(order)

('V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7')
('V1', 'V2', 'V3', 'V5', 'V4', 'V6', 'V7')
('V1', 'V3', 'V2', 'V4', 'V5', 'V6', 'V7')
('V1', 'V3', 'V2', 'V5', 'V4', 'V6', 'V7')
('V1', 'V3', 'V5', 'V2', 'V4', 'V6', 'V7')


4.1 ลำดับที่เป็นไปได้ทั้งหมด:

V1 V2 V3 V4 V5 V6 V7

V1 V2 V3 V5 V4 V6 V7

V1 V3 V2 V4 V5 V6 V7

V1 V3 V2 V5 V4 V6 V7

V1 V3 V5 V2 V4 V6 V7

4.2 ข้อมูล:[7, 3, 8, 1, 5, 9, 2]

ขั้นตอน:

[7, 3, 8, 1, 5, 9, 2]

[3, 7, 8, 1, 5, 9, 2]

[3, 7, 8, 1, 5, 9, 2]

[1, 3, 7, 8, 5, 9, 2]

[1, 3, 5, 7, 8, 9, 2]

[1, 3, 5, 7, 8, 9, 2]

[1, 2, 3, 5, 7, 8, 9]

4.3 จะไม่สามารถทำ Topological Sort ได้ เพราะมี cycle ในชีวิตจริง เช่น วิชา A ต้องผ่านวิชา B ก่อน แต่ B ก็ต้องผ่าน A ก่อน จึงลงทะเบียนไม่ได้

# ข้อ 5 ตรวจจับเหรียญปลอมในเครื่องหยอดเหรียญ (Fake-Coin + Binary Search)

In [ ]:
def fake_coin_group(n):
    count = 0

    while n > 1:
        n = (n + 2) // 3
        count += 1

    return count

print(fake_coin_group(27)) 
print(fake_coin_group(81)) 

3
4


5.1 แบ่งเป็น 3 กอง กองละ 9

ชั่งกอง A กับ B

- เท่ากัน → ปลอมอยู่ C
- ไม่เท่ากัน → ปลอมอยู่กองที่เบากว่า

จากนั้นทำซ้ำ: 27 → 9 → 3 → 1 จึงใช้ไม่เกิน 3 ครั้ง

5.2 81 เหรียญ: ceil(log₃81) = 4 

จึงใช้ 4 ครั้ง

5.3 1024 คน: log₂1024 = 10

การแบ่งครึ่งทำให้จำนวนข้อมูลลดลงครึ่งหนึ่งทุกครั้ง จึงใช้ O(log n)

# ข้อ 6 เกม Nim ในงานเลีย้งบริษัท (Variable-Size Decrease)

In [17]:
def nim_sum(piles):
    result = 0

    for pile in piles:
        result = result ^ pile

    return result


piles = [5, 3, 7]

result = nim_sum(piles)

print("Nim-sum =", result)

if result != 0:
    print("ผู้เล่นคนแรกชนะ")
else:
    print("ผู้เล่นคนที่สองชนะ")

Nim-sum = 1
ผู้เล่นคนแรกชนะ


6.1 5 XOR(⊕) 3 XOR(⊕) 7 = 1

เพราะ Nim-sum ≠ 0 ผู้เล่นแรกมี winning strategy

6.2 ผู้เล่นแรกหยิบ C ออก 4: (5,3,7) → (5,3,3)

Nim-sum: 5 XOR 3 XOR 3 = 5

ผู้เล่นที่สองต้องทำให้ Nim-sum = 0

ลด A จาก 5 → 0: (0,3,3)

0 XOR 3 XOR 3 = 0

ดังนั้นผู้เล่นที่สองควรหยิบ A ออก 5 ก้อน

6.3 Variable-Size-Decrease คือในแต่ละรอบขนาดปัญหาลดลงไม่เท่ากัน เช่น อาจหยิบ 1, 2, 3 หรือหลายก้อน

6.4 เริ่ม (5,3,7)

ผู้เล่นแรกสามารถทำให้ Nim-sum เป็น 0 เช่น:

C: 7 → 6

ได้: (5,3,6)

5 XOR 3 XOR 6 = 0

จากนั้นตอบโต้ให้ Nim-sum กลับเป็น 0 ทุกครั้ง

6.5 ถ้าผู้เล่นแรกเริ่มจากสถานะ (5,3,7) แล้วเดินแบบไม่ทำให้ Nim-sum เป็น 0 ผู้เล่นที่สองสามารถหาการเดินตอบโต้ที่ทำให้ Nim-sum กลับเป็น 0

# ข้อ 7 จัดเรียงธุรกรรมรายวัน (Mergesort vs Quicksort)

In [ ]:
# Merge Sort
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2

    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    result = []
    i = 0
    j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])

    return result


data = [950, 230, 780, 120, 560, 340, 890, 450, 670]
print(merge_sort(data))

[120, 230, 340, 450, 560, 670, 780, 890, 950]


In [19]:
# Quick Sort
def quick_sort(arr):
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]

    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort(left) + middle + quick_sort(right)


data = [950, 230, 780, 120, 560, 340, 890, 450, 670]
print(quick_sort(data))

[120, 230, 340, 450, 560, 670, 780, 890, 950]


7.1 แบ่งจนเหลือสมาชิกเดี่ยว แล้ว merge กลับ

ผลลัพธ์: [120, 230, 340, 450, 560, 670, 780, 890, 950]

Worst-case comparison สำหรับ n=9 สามารถนับจาก merge แต่ละระดับได้ประมาณ 19 ครั้ง

7.2 เลือก pivot เป็นตัวกลางของช่วง ถ้าข้อมูลเรียงแล้วและเลือกตัวกลางจริง ๆ การแบ่งยังค่อนข้างสมดุล จึงไม่จำเป็นต้องเกิด Worst Case

Worst Case ของ Quicksort เกิดเมื่อ partition ไม่สมดุล เช่น ขนาด: n−1 และ 0

T(n)=T(n−1)+Θ(n)

จึงได้: Θ(n²)

7.3 ถ้าต้องการรับประกันประสิทธิภาพทุกกรณี เลือก Mergesort เพราะ Θ(n log n) ทุกกรณี

# ข้อ 8 โครงสร้างองค์กรบริษัทขายตรง (Binary Tree Traversal)

In [ ]:
class Node:
    def __init__(self, value, sales):
        self.value = value
        self.sales = sales
        self.left = None
        self.right = None

def preorder(root):
    if root is None:
        return

    print(root.value, end=" ")
    preorder(root.left)
    preorder(root.right)

def inorder(root):
    if root is None:
        return

    inorder(root.left)
    print(root.value, end=" ")
    inorder(root.right)

def postorder_sum(root):
    if root is None:
        return 0

    left_sum = postorder_sum(root.left)
    right_sum = postorder_sum(root.right)

    return left_sum + right_sum + root.sales

root = Node("ก", 5)

root.left = Node("ข", 8)
root.right = Node("ค", 6)

root.left.left = Node("ง", 12)
root.left.right = Node("จ", 9)

root.right.left = Node("ฉ", 7)
root.right.right = Node("ช", 15)


print("Preorder:", end=" ")
preorder(root)
print()

print("Inorder:", end=" ")
inorder(root)
print()

print("Postorder:", end=" ")
def postorder(root):
    if root is None:
        return

    postorder(root.left)
    postorder(root.right)
    print(root.value, end=" ")

postorder(root)
print()

print("ยอดขายรวม =", postorder_sum(root), "ล้านบาท")

n = 7
A = n
C = 2 * n + 1
external_nodes = n + 1

print("จำนวนการบวก A(n) =", A)
print("จำนวนการเปรียบเทียบ C(n) =", C)
print("จำนวน External Nodes =", external_nodes)

# ข้อ 9 หาสาขาธนาคารที่ใกล้กันที่สุด (Closest-Pair by Divide-and-Conquer)

In [23]:
from math import sqrt

points = [
    (1, 2), (3, 7), (5, 1), (8, 6),
    (2, 9), (7, 3), (9, 8), (4, 5)
]

best = float("inf")
best_pair = None

for i in range(len(points)):
    for j in range(i + 1, len(points)):
        d = sqrt(
            (points[i][0] - points[j][0]) ** 2 +
            (points[i][1] - points[j][1]) ** 2
        )

        if d < best:
            best = d
            best_pair = (points[i], points[j])

print(best_pair)
print(best)

((3, 7), (2, 9))
2.23606797749979


9.1 จุด: S1(1,2)S2(3,7)S3(5,1)S4(8,6)S5(2,9)S6(7,3)S7(9,8)S8(4,5)

จำนวนคู่แบบ Brute Force: C(8,2)=28

คู่ที่ใกล้ที่สุด: S2-S8

ระยะ: sqrt((4-3)^2 + (5-7)^2)= sqrt(5) และ S4-S7 ก็มีระยะ sqrt(5)

ดังนั้น minimum distance = sqrt(5)

9.2 แบ่งตามแกน x เป็นสองส่วน

หา closest pair ในแต่ละฝั่งแล้วตรวจ strip ตรงกลาง

Recurrence: T(n)=2T(n/2)+Θ(n)

ดังนั้น:

a=2

b=2

d=1

ตาม Master Theorem: T(n)=Θ(n log n)

9.3 ถ้ามี 1 ล้านสาขา: Brute Force ≈ O(n²)

จะมีคู่ประมาณ: n(n−1)/2

ซึ่งใหญ่มาก

Divide-and-Conquer ลดเป็น Θ(n log n) จึงเหมาะกับข้อมูลขนาดใหญ่กว่า

# ข้อ 10 

In [ ]:
def gaussian_elimination(A, B):
    n = len(A)
    matrix = []

    for i in range(n):
        matrix.append(A[i] + [B[i]])

    for k in range(n - 1):

        for i in range(k + 1, n):
            factor = matrix[i][k] / matrix[k][k]

            for j in range(k, n + 1):
                matrix[i][j] -= factor * matrix[k][j]

    x = [0] * n

    for i in range(n - 1, -1, -1):
        total = 0

        for j in range(i + 1, n):
            total += matrix[i][j] * x[j]

        x[i] = (matrix[i][n] - total) / matrix[i][i]

    return x


A = [
    [2, -1, 1],
    [4, 1, -1],
    [1, 1, 1]
]

B = [1, 5, 0]


solution = gaussian_elimination(A, B)
print("x =", solution[0])
print("y =", solution[1])
print("z =", solution[2])

# ข้อ 11 จัดลำดับผู้ป่วยฉุกเฉิน (Heap Sort + AVL Tree)

In [24]:
def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


def build_max_heap(arr):
    n = len(arr)

    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)


data = [45, 88, 12, 95, 67, 34, 71, 53]

build_max_heap(data)

print(data)

[95, 88, 71, 53, 67, 34, 12, 45]


In [ ]:
# 11.1 สร้าง Max-Heap แบบ Bottom-Up ได้:

# [95, 88, 71, 53, 67, 34, 12, 45]

# Tree:

#           95
#         /    \
#       88      71
#      /  \    /  \
#    53   67  34  12
#   /
# 45

11.2 ถ้าเรียงจากน้อยไปมาก: [12, 34, 45, 53, 67, 71, 88, 95]

ถ้าจัดตาม "เร่งด่วนที่สุดก่อน": 95, 88, 71, 67, 53, 45, 34, 12

Time Complexity: Θ(n log n)

Basic operation: Sift Down / เปรียบเทียบ node กับลูก

11.3 เพิ่มท้าย: [95,88,71,53,67,34,12,45,80]

80 > 53 จึงสลับ: [95,88,71,80,67,34,12,45,53]

80 < 88 จึงหยุด

Heap ยังถูกต้อง

Heap เหมาะกับคิวฉุกเฉินเพราะสามารถดูผู้ป่วยที่มี Priority สูงสุดได้ที่ Root และการ Insert ใช้ O(log n)

# ข้อ 12 คำนวณดอกเบี้ย ทบต้น + ลดปัญหา (Horner's Rule + Problem Reduction)

In [1]:
def horner(P, x):
    result = P[-1]

    for i in range(len(P) - 2, -1, -1):
        result = result * x + P[i]

    return result


P = [10, 2, -7, 3, 5]

print(horner(P, 4))

1378


12.1 Nested form: P(t)=((((5t+3)t−7)t+2)t+10)

เมื่อ t=4: 

ขั้นที่ 1: 5(4)+3 = 23

ขั้นที่ 2: 23(4)−7 = 85

ขั้นที่ 3: 85(4)+2 = 342

ขั้นที่ 4: 342(4)+10 = 1378

ดังนั้น: P(4)=1378

จำนวนการคูณ = 4

จำนวนการบวก/ลบ = 4

Time Complexity: O(n)

Basic operation = การคูณและการบวกแต่ละรอบ

12.2 หา: 2¹³

13 ในฐานสอง: 1101₂

Left-to-Right: เริ่ม result = 1

bit 1:1² × 2 = 2

bit 1:2² × 2 = 8

bit 0:8² = 64

bit 1:64² × 2 = 8192

ดังนั้น: 2¹³ = 8192

Recurrence: T(n)=T(n/2)+Θ(1)

ดังนั้น:

a=1

b=2

d=0

Time Complexity: O(log n)

## 12.3

In [2]:
def gcd(a, b):
    while b != 0:
        a, b = b, a % b

    return a


def lcm(a, b):
    return (a * b) // gcd(a, b)


print(gcd(24, 60))
print(lcm(24, 60))

12
120


การ Reduction ช่วยเปลี่ยนปัญหา LCM ให้เป็นปัญหา GCD ซึ่งมี Algorithm ที่มีประสิทธิภาพ